synapse/http/matrixfederationclient.py

```diff
@@ -148,13 +179,29 @@ def get_json(self) -> Optional[JsonDict]:
         return self.json
 
 
+class JsonParser(ByteParser[Union[JsonDict, list]]):
+    """A parser that buffers the response and tries and parses it as JSON."""
+
+    CONTENT_TYPE = "application/json"
+
+    def __init__(self):
+        self._buffer = StringIO()
+        self._binary_wrapper = BinaryIOWrapper(self._buffer)
+
+    def write(self, data: bytes) -> int:
+        return self._binary_wrapper.write(data)
+
+    def finish(self) -> Union[JsonDict, list]:
+        return json_decoder.decode(self._buffer.getvalue())
+
+
 async def _handle_json_response(
```

Comment:
given this is now only called in three places, maybe it would be better just to inline it (ie, replace it with `_handle_response`)